Create a fuzz test that randomizes topology changes, cluster actions, and CQL operations

dcapwell opened the pull request.

dcapwell committed:
Create a fuzz test that randomizes topology changes, cluster actions, and CQL operations

dcapwell commented on test/distributed/org/apache/cassandra/distributed/impl/INodeProvisionStrategy.java:
```diff
@@ -29,137 +29,80 @@
 {
-    enum Strategy
+    @FunctionalInterface
+    interface Factory
```
this lets me control this interface a lot more in my test... right now the code assumes node1 never goes away, but my test invalidates that assumption... so seed selection and more must live with my test

dcapwell commented on test/unit/accord/utils/Gen.java:
```diff
@@ -87,6 +89,22 @@ default Gen<A> filter(Predicate<A> fn)
         };
     }
 
+    default Gen<A> filter(int maxAttempts, A defaultValue, Predicate<A> fn)
```
a lot of changes to these classes are not *needed* for this patch, but my thinking is to try to make in-sync with accord so the next rebase isn't a pain

dcapwell commented on test/unit/accord/utils/Gens.java:
```diff
@@ -19,26 +19,39 @@
 package accord.utils;
 
 import java.lang.reflect.Array;
+import java.math.BigDecimal;
```
a lot of changes to these classes are not *needed* for this patch, but my thinking is to try to make in-sync with accord so the next rebase isn't a pain

dcapwell commented on test/unit/accord/utils/Gens.java:
```diff
@@ -89,6 +145,215 @@ public static <T> Gen<T> pick(Map<T, Integer> values)
+            indexes = null;
+        }
+        return rs -> {
+            switch (rs.nextInt(0, 2))
```
this isn't 100% in-sync with accord due to a missing set of random methods I didn't care to backport... so this is missing 2 other "distributions"

dcapwell commented on test/unit/accord/utils/Gens.java:
```diff
@@ -433,16 +735,7 @@ public Gen<T[]> ofSizeBetween(int minSize, int maxSize) {
-                        return Arrays.copyOf(list, i);
-                    }
-                }
+                    list[i] = fn.next(r);
```
I don't know when/why this was removed and looking at the `List` version this looks like a bug... we don't hit this case often with arrays which is prob why we didn't notice...

I plan to create a patch for cep-15-accord + accord to keep them in-sync with this PR... ill try to mixup in both places soon

dcapwell commented on test/unit/accord/utils/Property.java:
```diff
@@ -18,23 +18,25 @@
 {
     public static abstract class Common<T extends Common<T>>
     {
-        protected long seed = ThreadLocalRandom.current().nextLong();
```
`ThreadLocalRandom` should be avoided at all costs... the seed isn't random which means restarting the JVM yields the same seed... so you keep running the same tests over and over and over again...

dcapwell commented on test/unit/org/apache/cassandra/utils/ConfigGenBuilder.java:
```diff
@@ -70,7 +70,7 @@ public enum Memtable
      * When loading the {@link Config} from a yaml its possible that some configs set will conflict with the configs that get generated here, to avoid that set them to a good default
      */
-    public static Config santize(Config config)
+    public static Config sanitize(Config config)
```
@maedhroz you missed a typo!  SHAME!!!!!

dcapwell committed:
jvm dtest improvements to have better error msg.  harry fixes to not query downed nodes

dcapwell committed:
stdize the postfix for each command which made it easier to add cms members to that as well

dcapwell committed:
document why write replicas is used for the test

dcapwell committed:
ignore cms snapshot voting error

dcapwell committed:
fixed one more place where the coordinator is assumed to be first

dcapwell commented on src/java/org/apache/cassandra/tcm/AbstractLocalProcessor.java:
```diff
@@ -56,7 +56,7 @@ public final Commit.Result commit(Entry.Id entryId, Transformation transform, fi
             ClusterMetadata previous = log.waitForHighestConsecutive();
             if (!previous.fullCMSMembers().contains(FBUtilities.getBroadcastAddressAndPort()))
-                throw new IllegalStateException("Node is not a member of CMS anymore");
+                throw new IllegalStateException("Node " + FBUtilities.getBroadcastAddressAndPort() + " is not a member of CMS anymore in " + previous.epoch + "; members=" + previous.fullCMSMembers());
```
this was super flakey and added a ignore filter so the test won't fail if this is seen anymore (assuming its safe... think it is)...

I saw a history where node1 starts a snapshot, node1 looses its CMS membership (nodes 2/3/4 are now the members), then this check blows up...

The extra details just help know *when* in the timeline things happened...

dcapwell replied:
I am going to drop as Marcus is making similar change in his patch....

dcapwell replied:
https://issues.apache.org/jira/browse/CASSANDRA-19872

dcapwell committed:
cleanup

dcapwell committed:
imports

dcapwell committed:
added hook so harry can do one last read to validate all data if anything exists

dcapwell commented on test/distributed/org/apache/cassandra/fuzz/topology/HarryTopologyMixupTest.java:
```diff
@@ -0,0 +1,159 @@
+        @Override
+        protected void onConfigure(IInstanceConfig config)
+        {
+            config.set("metadata_snapshot_frequency", 5);
```
I don't remember why I changed this... I have yet to find a history for *why*

dcapwell committed:
disable partitioner support in config gen rather than hard code murmur again

dcapwell committed:
document how to avoid bind failures on mac

dcapwell committed:
move the remove before the set in case the yaml generator ever wants to support legacy configs

dcapwell commented on test/distributed/org/apache/cassandra/fuzz/topology/TopologyMixupTestBase.java:
```diff
@@ -0,0 +1,761 @@
+
+        protected void onStartupComplete(long tcmEpoch)
+        {
+
```
not used in this patch but accord needs to block on accord knowing this epoch

dcapwell committed:
forgot to commit imporant classes

dcapwell committed:
doubling steps as the test only takes 4m with the update so covers more space

dcapwell committed:
revert allowing the CMS membership loss error as Marcus fixed that.  Before each action also check if the cluster has seen an error and fail fast

dcapwell committed:
refactor so the precheck for harry is shared and made simple command allow a name function rather than just a string

dcapwell committed:
do reads 20% of the time

ifesdjeen commented on src/java/org/apache/cassandra/tcm/ClusterMetadataService.java:
```diff
@@ -844,11 +844,22 @@ private Pair<State, Processor> delegateInternal()
+                    retryPolicy.maybeSleep();
+                }
+            }
+            return Commit.Result.failed(ExceptionCode.SERVER_ERROR, "Could not commit " + transform.kind() + " at epoch " + lastKnown);
```
nit: should we use string format here?

dcapwell replied:
I believe im reverting this as Marcus is pushing a fix in his own JIRA

ifesdjeen commented on src/java/org/apache/cassandra/tcm/ClusterMetadataService.java:
```diff
@@ -844,11 +844,22 @@ private Pair<State, Processor> delegateInternal()
-            if (delegate.left == LOCAL || delegate.left == RESET)
-                replicator.send(result, null);
-            return result;
+            while (!retryPolicy.reachedMax())
```
I _think_ we should not retry here, and instead should retry in an underlying processor. But maybe this is one of the Marcus's patches.

dcapwell replied:
I believe im reverting this as Marcus is pushing a fix in his own JIRA

ifesdjeen commented on test/distributed/org/apache/cassandra/distributed/shared/ClusterUtils.java:
```diff
@@ -611,8 +632,8 @@ public static void waitForCMSToQuiesce(ICluster<IInvokableInstance> cluster, Epo
                 Epoch version = getClusterMetadataVersion(cluster.get(j));
-                if (!awaitedEpoch.equals(version))
-                    notMatching.add(new ClusterMetadataVersion(j, version, getClusterMetadataVersion(cluster.get(j))));
+                if (!awaitedEpoch.equals(version) && version.getEpoch() < awaitedEpoch.getEpoch())
```
Just `version.getEpoch() < awaitedEpoch.getEpoch()` should be enough here, or?

ifesdjeen commented on test/distributed/org/apache/cassandra/fuzz/topology/HarryTopologyMixupTest.java:
```diff
@@ -0,0 +1,160 @@
+                harryState.generation = history.generation();
+                return validateAll;
+            }
+            if (harryState.numInserts > 500
```
I do not mind the current state, but am wondering if a slightly different interface might be more convenient? One option would be to use something like "stateful model checker" from QT, where you would specify the possible commands, and supply their probabilities (i.e. 0.8 / 0.2 writes to reads distribution), something like a `RangeTombstoneBurnTest` would do:

```
var insertChance = 0.8;
var readChacne = 0.2;
                            .step((history, rng) -> rng.nextFloat() > insertChance,
                                  (history, rng) -> {
...
                                  })
                            .step((history, rng) -> rng.nextFloat() > readChance,
                                  (history, rng) -> {
..
                                  })
```

This interface also nicely fits with the `State` idea you got here.

dcapwell replied:
>  like "stateful model checker" from QT

where is this?  your example is harry's and the QT one Jordan worked on never merged nor really got used in the project.

dcapwell replied:
your comment is a "builder" of `Command`?  In this case I don't feel it makes too much sense as we have another thing to address

```
            if (harryState.generation != history.generation())
            {
                harryState.generation = history.generation();
                return validateAll;
            }
```

right now the logic says 20% of the time we do reads *or* we have done 500 inserts (honestly copy/paste from harry code... no issues dropping ^_^).

TL;DR - this `Command` does

1) if a topology change happened, always do a read
2) 20% of the time do a read
3) 80% of the time do a write

I am not sure how a builder would make this easier to read...

ifesdjeen commented on test/distributed/org/apache/cassandra/fuzz/topology/TopologyMixupTestBase.java:
```diff
@@ -0,0 +1,756 @@
+            }
+            preActions.add(new Runnable()
+            {
+                // in order to remove this action, an annonomus class is needed so "this" works, lambda "this" is the parent class
```
annonomus > anonymous

ifesdjeen replied:
Also, if you like you can use `preActions.remove(State.this);` and lambda. I don't mind an anonymous class though

dcapwell replied:
> if you like you can use preActions.remove(State.this);

what do you mean?  if this becomes a lambda then `this` is `State.this` which is not the correct reference to remove... I would need to save to a variable to keep access *or* make it anonymous...

ifesdjeen commented on test/distributed/org/apache/cassandra/fuzz/topology/TopologyMixupTestBase.java:
```diff
@@ -0,0 +1,756 @@
+
+        public int[] active()
+        {
+            int[] keys = new int[nodes.size()];
```
this one seems to be returning all nodes not just active, or?

dcapwell replied:
dead code... this *used* to be different but I removed its usage... just didn't cleanup

ifesdjeen commented on test/distributed/org/apache/cassandra/fuzz/topology/TopologyMixupTestBase.java:
```diff
@@ -0,0 +1,756 @@
+        public Node addNode()
+        {
+            int id = ++uniqueInstances;
+            List<String> instTokens = Gens.lists(MURMUR_TOKEN_GEN
```
should we make this static? looks like we want just one instance of this generator, or?

dcapwell replied:
not possible to be static.

the first line is

`Gens.lists(MURMUR_TOKEN_GEN.filterAsInt(t -> !activeTokens.contains(Integer.toString(t))))`

this references *current* state, which can't be accessed statically.

second

`.unique()` "could" be static but isn't thread safe... so dangerous to do.

ifesdjeen reviewed (commented):
I like the direction where this is headed and am mostly on board with changes. The only thing I wanted to propose is a possible modification in the API. Rather than pre-declaring possible commands that could potentially be executed, we could do something like:

```
                            .command((state, rng) -> {
                                      rng.nextFloat() > threshold && // compute chance
                                      predicate.test(state)          // precondition (for example, do we have enough node to even start decomission 
                                  },
                                  (state, rng) -> {
                                      return multistep(...
                                  })
                            .command((state, rng) -> {
                                      rng.nextFloat() > threshold && // compute chance
                                      predicate.test(state)          // precondition (for example, do we have enough node to even start decomission 
                                  },
                                  (state, rng) -> {
                                      return multistep(...
                                  })
```

WDYT?

dcapwell committed:
Revert "patch by marcus to fix tcm issues found"

This reverts commit 4bab6145505d1edec796a1632d178b4ff9d56c10.

dcapwell committed:
adding new builder interface

dcapwell committed:
remove the over 500 write thing and just do 20% reads

dcapwell committed:
typo

dcapwell committed:
simplify boolean condition, the RHS already covers the LHS

dcapwell committed:
remove dead code

dcapwell commented:
fixed all feedback or asked for more details (also poked you in slack)

you have 2 similar / related feedbacks, so trying to address here (GH isn't letting me reply...)

https://github.com/apache/cassandra/pull/3486#pullrequestreview-2263041945 - I tested out a `CommandsBuilder` and it made the test `RouteIndextTest` in cep-15-accord easier to read, but I don't see how it helps in this case due to 2 variables that I am not sure how to handle (without making the builder specific to this class)

```
final List<Runnable> preActions = new CopyOnWriteArrayList<>();
final List<Runnable> onError = new CopyOnWriteArrayList<>();
```

The `State` has a `preActions` and `onError` actions... these kinda distort the model and are plugged into the `Commands` functions... if I move to a `CommandsBuilder` model I think `onError` doesn't really change much, but `preActions` gets funky.  Personally I feel the `preActions` is a hack... it would be great to return `Command` so the history does reflect those steps... but that would kinda require each `Command` returned to be a `MultiCommand`... and what about actions that don't mutate anything and just update the `State`'s bookkeeping?  those don't change the history and would just clutter up the error message...


https://github.com/apache/cassandra/pull/3486#discussion_r1732701880

I don't see how a builder works in this specific case.  We are creating a single `Command` and this command has 3 behaviors: if topology changed do read, 20% of the time do reads, 80% of the time do writes.... the builder would need to handle the `if topology changed` bit... which seems funky... and doing a builder for the 20%/80% case seems more verbose than the current logic

dcapwell committed:
switched to CommandsBuilder

dcapwell commented:
ok pushed to using the new `CommandsBuilder`... the test is now

```
statefulBuilder.check(commands(this::stateGen)
                              .preCommands(state -> state.preActions.forEach(Runnable::run))
                              .add(2, (rs, state) -> {
                                  EnumSet<TopologyChange> possibleTopologyChanges = possibleTopologyChanges(state);
                                  if (possibleTopologyChanges.isEmpty()) return ignoreCommand();
                                  return topologyCommand(state, possibleTopologyChanges).next(rs);
                              })
                              .add(1, (rs, state) -> repairCommand(state, rs.pickInt(state.topologyHistory.up())))
                              .add(7, (rs, state) -> state.statementGen.apply(rs, state))
                              .destroyState((state, cause) -> {
                                  TopologyMixupTestBase.this.destroyState(state, cause);
                                  state.close();
                              })
                              .build());
```

still not a fan of `preCommands` but also not a fan of what I am doing now... at least this interface doesn't make things worse...

one thing that is nice with this builder is I flip `Gen<Command>, Weight`, I personally prefer weight first then the command, always found the gen first harder to read (its an impl detail, we store the Gen in the map as the key)

dcapwell commented:
also removed `onError` as that is an accord thing... I added that for better debugging of timeout... it doesn't need to be in the base interface and can be done in `AccordTopologyMixupTest.destroyState`, so moving it there

dcapwell committed:
we have java11 can do try (state) to avoid having to miss cleaning up the state

dcapwell committed:
import

dcapwell committed:
cleanup

dcapwell committed:
cleanup

dcapwell committed:
added new add function to support random weights

dcapwell committed:
fixed bug where the weights are stateful, causing a rerun of a failed seed to not work

ifesdjeen commented on test/distributed/org/apache/cassandra/fuzz/topology/TopologyMixupTestBase.java:
```diff
@@ -0,0 +1,697 @@
+                                 .withConfig(c -> {
+                                     c.with(Feature.values())
+                                      .set("write_request_timeout", "10s");
+                                     //TODO (mainatiance): where to put this?  Anything touching ConfigGenBuilder with jvm-dtest needs this...
```
Nit: mainatiance (and maintnance) below :)

ifesdjeen commented on test/distributed/org/apache/cassandra/fuzz/topology/TopologyMixupTestBase.java:
```diff
@@ -0,0 +1,697 @@
+                                 //TODO (maintance): AbstractCluster's Map<Integer, NetworkTopology.DcAndRack> nodeIdTopology makes playing with dc/rack annoying, if this becomes an interface then TopologyHistory could own
+                                 .withNodeProvisionStrategy((subnet, portMap) -> new INodeProvisionStrategy.AbstractNodeProvisionStrategy(portMap)
+                                 {
+                                     {
```
nit: probably extra brackets are not necessary

dcapwell replied:
in this case not true.  This is the constructor

ifesdjeen commented on test/unit/accord/utils/async/TimeoutUtils.java:
```diff
@@ -0,0 +1,70 @@
+
+    public static void runBlocking(Duration timeout, String threadName, FailingRunnable fn) throws ExecutionException, InterruptedException, TimeoutException
+    {
+        // MAINTAINCE: Once the accord branch merges to trunk this can be dropped and will be AsyncChain again, but since this is forked into C* (that doesn't have AsyncChain) need to use Futures
```
nit: MAINTAINCE

dcapwell committed:
feedback

dcapwell committed:
feedback

dcapwell committed:
feedback

dcapwell committed:
rename to match trunk

dcapwell committed:
went back to 20 steps as we dont have a good long running build target yet

dcapwell committed:
added new methods cep-15-accord wanted

dcapwell committed:
bad rebase
